tests/tests/many_threads.c

```diff
@@ -1,51 +1,83 @@
-// ignore: broken during new control point design
-// Compiler:
 // Run-time:
+//   env-var: YKD_PRINT_IR=jit-pre-opt,aot
+//   env-var: YKD_SERIALISE_COMPILATION=1
+//   env-var: YKD_PRINT_JITSTATE=1
+//   stderr:
+//     ...
+//     jit-state: enter-jit-code
+//     ...
 
 // Check that compiling and running traces in parallel works.
+//
+// Due to non-determinism, it's difficult to devise a meaningful expected
+// outcome. Although threads will block waiting for the trace compiler (due to
+// YKD_SERIALISE_COMPILATION=1), there is still non-determinism in which
+// threads execute compiled trace code and how they might interleave.
+//
+// For this reason, this test only really checks that nothing crashes.
 
 #include <assert.h>
 #include <err.h>
 #include <pthread.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <yk.h>
 #include <yk_testing.h>
 
 #ifdef linux
 #include <sys/sysinfo.h>
 #endif
 
-static void *trace(void *unused) {
-  for (int i = 0; i < 3; i++) {
-    __yktrace_start_tracing(HW_TRACING);
-    int res = 1 + 1;
-    void *tr = __yktrace_stop_tracing();
-    assert(res == 2);
-
-    void *ptr = __yktrace_irtrace_compile(tr);
-    __yktrace_drop_irtrace(tr);
-    void (*func)() = (void (*)())ptr;
-    func();
+#define ITERS 1000
+
+struct thread_data {
+  int tnum;
+  YkLocation *loc;
+  YkMT *mt;
+};
+
+static void *trace(void *arg) {
+  struct thread_data *td = (struct thread_data *)arg;
+
+  uintptr_t i = ITERS;
+  NOOPT_VAL(i);
+  while (i > 0) {
+    yk_mt_control_point(td->mt, td->loc);
+    fprintf(stderr, "thread #%d: %lu\n", td->tnum, i);
```

Comment:
We can't do this on every iteration as it will synchronise the threads. Ideally we wouldn't print out anything during the main loop at all: if we need to print something we should only print every *n* iterations (where *n* would be quite a high number).

Reply:
In fact, since we don't use this output in the test at all, can't we just get rid of this `printf` entirely?

Reply:
Why would it synchronise threads? I saw interleaving in my testing.

Reply:
e.g.:
```
jit-state: enter-jit-code
thread #24: 91591
jit-state: enter-jit-code
thread #9: 91338
jit-state: exit-jit-code
jit-state: enter-jit-code
jit-state: thread #9: 91337
exit-jit-code
jit-state: enter-jit-code
thread #15: 91362
jit-state: exit-jit-code
jit-state: enter-jit-code
jit-state: enter-jit-code
thread #16: 91499
jit-state: exit-jit-code
thread #26: 91629
jit-state: exit-jit-code
jit-state: exit-jit-code
jit-state: exit-jit-code
jit-state: exit-jit-code
jit-state: exit-jit-code
jit-state: exit-jit-code
jit-state: exit-jit-code
jit-state: exit-jit-code
jit-state: enter-jit-code
jit-state: thread #25: 91467
enter-jit-code
thread #13: 91326
jit-state: exit-jit-code
```

Reply:
It's hard to say anything with certainty, but `printf` and friends have to gain a lock; that means at most one thread will `print` at any point, and threads will probably put themselves to sleep on most iterations of the loop. This makes the test a lot less nondeterministic than we would like.

Reply:
Perhaps only one thread can printf at a time, but the log above clearly shows that the rest of the trace code is running with all kinds of interleaving.

Reply:
There's no "perhaps" about it ;) You're interpreting what I say as "the `printf` means there is no interleaving" but that is not what I'm saying: what I'm saying is that the `printf` reduces nondeterminism (note: "reduces" not "precludes"!). [Since we don't even use the `printf` output in the lang_tester test, the `printf` also makes the computer get hot without allowing us to test anything useful!]

It would be a better test for us if instead of a `printf` we did some sort of normal calculation (preferably one that the optimiser won't remove entirely) that doesn't do locking.

Reply:
>  some sort of normal calculation

We are already checking that teach thread decrements down to zero, and a `NOOPT_VAL` ensures the loop can't disappear.

I'll just kill the printf.

I find this test unsatisfying, but I don't think it's possible to write a satisfying one.

Reply:
Here's an example of something we could do: we can pass a number to each thread based on how many we've created already, and do some sort of calculation on that in the loop. Then we can check that against the output at the end of the thread.

Reply:
Admittedly I'm not sure what we can actually execute in the stopgap interpreter at the moment. Maybe `i--`  is as far as we should push our luck?

Reply:
True. We can still write the proper test now, and once Lukas merges his work, we will be testing the whole program as intended.

Reply:
Not sure I understand: are you saying we should to the test now or wait until the stopgap interpreter is merged?

Reply:
Every test in the C suite currently can only test up to the point of stop-gapping, where upon the program exits (with success status for now, otherwise we couldn't test anything).

This threading test, like the others, should test up until we stop-gap, but we can still put assertions in for when we can stop-gap.

Agreed?

Reply:
I still don't know if you're suggesting adding more stuff in or not :)

Reply:
In fact, it looks like some threads do manage to exit before the compiled trace is available, so we can actually check the result of some computation for those threads now without having stop-gapping.

Reply:
> I still don't know if you're suggesting adding more stuff in or not :)

probably best to wait for the diff. Bear with me because something is crashing.